quantum/backlight/backlight_arm.c

```diff
@@ -0,0 +1,261 @@
+#include "quantum.h"
+#include "backlight.h"
+#include <hal.h>
+#include "debug.h"
+
+// TODO: remove short term bodge when refactoring BACKLIGHT_CUSTOM_DRIVER out
+#ifdef BACKLIGHT_PIN 
+
+#if defined(STM32F1XX) || defined(STM32F1xx)
+    #define USE_GPIOV1
+#endif
+
+//GPIOV2 && GPIOV3
+#ifndef BACKLIGHT_PAL_MODE
+    #define BACKLIGHT_PAL_MODE      2
+#endif
+
+// GENERIC
+#ifndef BACKLIGHT_PWM_DRIVER
+    #define BACKLIGHT_PWM_DRIVER    PWMD3
+#endif
+#ifndef BACKLIGHT_PWM_CHANNEL
+    #define BACKLIGHT_PWM_CHANNEL    3
+#endif
+
+static void breathing_callback(PWMDriver *pwmp);
+
+static PWMConfig pwmCFG = {
+  0xFFFF,                              /* PWM clock frequency  */
+  256,                              /* PWM period (in ticks) 1S (1/10kHz=0.1mS 0.1ms*10000 ticks=1S) */
+  NULL,                               /* No Callback */
+  {
+      {PWM_OUTPUT_DISABLED, NULL}, /* Enable Channel 0 */
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL}
+  },
+  0,                                  /* HW dependent part.*/
+  0
+};
+
+static PWMConfig pwmCFG_breathing = {
+  0xFFFF,                              /* 10kHz PWM clock frequency  */
+  256,                              /* PWM period (in ticks) 1S (1/10kHz=0.1mS 0.1ms*10000 ticks=1S) */
+  breathing_callback,                               /* Breathing Callback */
+  {
+      {PWM_OUTPUT_DISABLED, NULL}, /* Enable Channel 0 */
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL}
+  },
+  0,                                  /* HW dependent part.*/
+  0
+};
+
+// See http://jared.geek.nz/2013/feb/linear-led-pwm
+static uint16_t cie_lightness(uint16_t v) {
```

Comment:
I keep seeing this function in PRs :| how many copies of this exist now?

Reply:
OK, RG says there are about 10 copies already, so I guess this is now a thing?

Reply:
Too many, its one of the tasks I intend to clean up during my first round of refactoring.